boards: st: nucleo_h7s3l8: Introduce App in Ext Flash variant

TD-JBL opened the pull request:
Similarly to what is done for stm32h7s78_dk board in #97992, introduce an ext_flash_app board variant aiming at being chainloaded by MCUBoot.
Requires mcuboot counter part (https://github.com/mcu-tools/mcuboot/pull/2510) to allow MCUBoot to run on internal flash.

Need / waiting for mcu-tools/mcuboot/pull/2554.

**Tested with**
* without sysbuild: `samples/hello_world` with `nucleo_h7s3l8` (app in internal flash)
* with sysbuild: `samples/hello_world` with `nucleo_h7s3l8/stm32h7s3xx/ext_flash_app` (mcuboot + app in external flash)
* with sysbuild: `samples/sysbuild/with_mcuboot` with `nucleo_h7s3l8/stm32h7s3xx/ext_flash_app` (mcuboot + app in external flash)

TD-JBL committed:
boards: arm: st: nucleo_h7s3l8: Add *-block-size properties to xspi2

Add write-block-size and erase-block-size attributes to xspi nor flash
connected to xspi2. This is needed for mcuboot to handle flash access
of the ext_flash. The changes in the node structure is taken from the
stm32h7s78_dk board.

Update zephyr,flash-controller in samples/sysbuild/with_mcuboot/boards
/nucleo_h7s3l8.overlay to have CONFIG_STM32_APP_IN_EXT_FLASH set when
direct parent of flash is not XSPI but only the parent of the flash-
controller.

Signed-off-by: Thomas Decker <decker@jb-lighting.de>

TD-JBL committed:
boards: st: nucleo_h7s3l8: Add ext_flash_app variant

Add ext_flash_app variant to nucleo_h7s3l8 to allow building
applications running on ext flash.

Based on Erwan Gouriou's work on stm32h7s78_dk in PR #97992
(which is based on A.Jarmouni's work on stm32h750_dk in PR 97037).

Signed-off-by: Thomas Decker <decker@jb-lighting.de>

TD-JBL committed:
samples: sysbuild: with_mcuboot: nucleo_h7s3l8: Use ext_flash_app var…

…iant

Now that ext_flash_app variant is available, it should be used for
mcuboot related apps on this board.

Delete boards/nucleo_h7s3l8.overlay (and boards/stm32h7s78_dk.overlay)
which are not needed anymore.

Signed-off-by: Thomas Decker <decker@jb-lighting.de>

nordicjm commented on boards/st/nucleo_h7s3l8/Kconfig.sysbuild:
```diff
@@ -0,0 +1,19 @@
+endchoice
+
+choice MCUBOOT_MODE
+	default MCUBOOT_MODE_SINGLE_APP
```
direct xip?

nordicjm replied:
they are different modes, single slot means you use slot 0 and cannot do any updates from the app, MCUboot must have serial recovery mode to update slot 0, with direct-xip you can load firmware updates to the opposing slot and MCUboot will boot the highest version

erwango reviewed (changes requested):
Looks ok, but please also update board documentation (you'll find examples for other boards)

TD-JBL commented:
Added documentation (copied and adapted from stm32h750b_dk). Also changed to `MCUBOOT_MODE_SWAP_USING_OFFSET`.

TD-JBL commented:
rebased because of merge conflicts in samples/sysbuild/with_mcuboot/sample.yaml

etienne-lms commented on boards/st/nucleo_h7s3l8/Kconfig.sysbuild:
```diff
@@ -0,0 +1,19 @@
+
+choice MCUBOOT_MODE
+	default MCUBOOT_MODE_SWAP_USING_OFFSET
+endchoice
```
Recently merged P-R https://github.com/zephyrproject-rtos/zephyr/pull/100385 made swap-using-offset the default MCUBoot mode for STM32 targets.
I think this directive can be removed.

etienne-lms commented on boards/st/nucleo_h7s3l8/nucleo_h7s3l8_stm32h7s3xx_ext_flash_app.dts:
```diff
@@ -0,0 +1,54 @@
+};
+
+&ext_flash {
+	status = "okay";
```
`status = <okay>` is not required.

TD-JBL replied:
Removed

etienne-lms reviewed (commented):
Looks good to me with minor comments to address or not.

TD-JBL commented on boards/st/nucleo_h7s3l8/doc/index.rst:
```diff
@@ -213,6 +213,7 @@ the ``--runner`` (or ``-r``) option:
 .. code-block:: console
 
    $ west flash --runner openocd
+   $ west flash --runner jlink
```
Sure, I will remove this line

TD-JBL committed:
boards: st: stm32h573i_dk & stm32h750b_dk: Remove MCUBOOT_MODE

Remove MCUBOOT_MODE_SWAP_USING_OFFSET from the Kconfig.sysbuild of the
boards stm32h573i_dk and stm32h750b_dk, because its the default
MCUBOOT_MODE since #100385.

Signed-off-by: Thomas Decker <decker@jb-lighting.de>

JarmouniA commented on boards/st/nucleo_h7s3l8/Kconfig.sysbuild:
```diff
@@ -12,8 +12,4 @@ choice BOOT_SIGNATURE_TYPE
 	default BOOT_SIGNATURE_TYPE_NONE
 endchoice
 
-choice MCUBOOT_MODE
```
Should be removed directly from the commit where it's added.

JarmouniA commented on boards/st/stm32h573i_dk/Kconfig.sysbuild:
```diff
@@ -1,6 +1,8 @@
 # SPDX-FileCopyrightText: Copyright The Zephyr Project Contributors
 # SPDX-License-Identifier: Apache-2.0
 
+if BOARD_STM32H573I_DK_STM32H573XX_EXT_FLASH_APP
```
No, mcuboot can be used without ext Flash as well on this board.
